src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1304 @@
+
+        for i in 0..node_data.len() {
+            // dump the node to storage
+            write_nodetype_bytes(f, &node_data[i].0, node_data[i].1)?;
+            
+            // next node
+            fseek(f, offsets[i] as u64)?;
+        }
+
+        Ok(ptr)
+    }
+
+    /// Dump ourself to f
+    pub fn dump<F: Read + Write + Seek>(&mut self, f: &mut F, bhh: &BlockHeaderHash, parent_bhh: &BlockHeaderHash) -> Result<u64, Error> {
+        if self.block_header == *bhh {
+            let root_ptr = self.root_ptr();
+            let (root, hash) = self.read_node(&TriePtr::new(TrieNodeID::Node256, 0, root_ptr as u32))?;
+            self.dump_traverse(f, &root, &hash, parent_bhh)
+        }
+        else {
+            trace!("Failed to dump {:?}: not the current block", bhh);
+            Err(Error::NotFoundError)
+        }
+    }
+
+    fn size_hint(&self) -> usize {
+        self.total_bytes
+    }
+}
+
+impl TrieStorage for TrieRAM {
+    fn extend(&mut self, bhh: &BlockHeaderHash) -> Result<(), Error> {
+        if self.block_header == *bhh {
+            return Err(Error::ExistsError);
+        }
+        trace!("Extend to {:?}", bhh);
+        self.block_header = bhh.clone();
+        self.offset = 0;
+        self.num_nodes = 0;
+        self.data.clear();
+        self.readonly = false;
+        Ok(())
+    }
+
+    fn open(&mut self, bhh: &BlockHeaderHash, readwrite: bool) -> Result<(), Error> {
+        if self.block_header != *bhh {
+            trace!("Failed to open {:?}: not the current block", bhh);
+            return Err(Error::NotFoundError);
+        }
+        self.block_header = bhh.clone();
+        self.offset = 0;
+        self.num_nodes = self.data.len() as u64;
+        self.readonly = !readwrite;
+        Ok(())
+    }
+
+    fn tell(&self) -> BlockHeaderHash {
+        self.block_header.clone()
+    }
+    
+    fn block_walk(&mut self, _back_block: u32) -> Result<BlockHeaderHash, Error> {
+        panic!("Not implemented for TrieRAM");
+    }
+
+    fn root_ptr(&self) -> u64 { 0 }
+
+    fn readwrite(&self) -> bool {
+        !self.readonly
+    }
+
+    fn format(&mut self) -> Result<(), Error> {
+        if self.readonly {
+            trace!("Read-only!");
+            return Err(Error::ReadOnlyError);
+        }
+
+        self.data.clear();
+        self.offset = 0;
+        self.num_nodes = 0;
+        Ok(())
+    }
+
+    fn read_node_hash_bytes(&mut self, ptr: &TriePtr, buf: &mut Vec<u8>) -> Result<(), Error> {
+        if (ptr.ptr() as u64) >= (self.data.len() as u64) {
+            trace!("TrieRAM: Failed to read node bytes: {} >= {}", ptr.ptr(), self.data.len());
+            Err(Error::NotFoundError)
+        }
+        else {
+            fast_extend_from_slice(buf, self.data[ptr.ptr() as usize].1.as_bytes());
+            Ok(())
+        }
+    }
+
+    fn read_node(&mut self, ptr: &TriePtr) -> Result<(TrieNodeType, TrieHash), Error> {
+        let disk_ptr = ftell(self)?;
```

Comment:
Is `let disk_ptr = ftell(self)?;` equivalent to `let disk_ptr = self.offset;` ?

Reply:
Yes -- `ftell()` takes something that has the `Seek` trait, so internally, you're getting back the value of `self.offset`.

Reply:
This should be addressed -- storage is no longer seek-able.